sycl/doc/extensions/proposed/sycl_ext_oneapi_annotated_ptr_properties.asciidoc

```diff
@@ -0,0 +1,158 @@
+
+== Introduction
+This extension defines the properties supported with the class `annotated_ptr`
+which is defined by the extension `sycl_ext_oneapi_annotated_ptr`. These
+properties are carried by the `annotated_ptr` class and implemented by the
+toolchain.
+
+== Notice
+
+Copyright (c) 2021 Intel Corporation.  All rights reserved.
+
+NOTE: Khronos(R) is a registered trademark and SYCL(TM) and SPIR(TM) are
+trademarks of The Khronos Group Inc.  OpenCL(TM) is a trademark of Apple Inc.
+used by permission by Khronos.
+
+== Status
+
+Final Draft
+
+This is a preview extension specification, intended to provide early access to
+a feature for review and community feedback. When the feature matures, this
+specification may be released as a formal extension.
+
+Because the interfaces defined by this specification are not final and are
+subject to change they are not intended to be used by shipping software
+products.
+
+== Version
+
+Built On: {docdate} +
+Revision: 1
+
+== Contact
+
+Abhishek Tiwari, Intel (abhishek2 'dot' tiwari 'at' intel 'dot' com)
+
+== Contributors
+
+Abhishek Tiwari, Intel +
+Joseph Garvey, Intel
+
+
+== Dependencies
+
+This extension is written against the SYCL 2020 specification, revision 4.
+
+It depends on the following extensions:
+ - sycl_ext_oneapi_annotated_ptr
+ - sycl_ext_intel_buffer_location
+
+== Overview
+
+This extension introduces properties that establish the annotations that will be
+carried by the class `sycl::ext::oneapi::annotated_ptr`.
+
+An example of the syntax can be seen below. 
+
+[source,c++]
+----
+annotated_ptr<MyClass, property_list_t<align<16>>> aptr;
+----
+
+== Feature test macro
+
+This extension provides a feature-test macro as described in the core SYCL
+specification section 6.3.3 "Feature test macros". Therefore, an
+implementation supporting this extension must predefine the macro
+`SYCL_EXT_ONEAPI_ANNOTATED_PTR_PROPERTIES` to one of the values defined in the
+table below.
+Applications can test for the existence of this macro to determine if the
+implementation supports this feature, or applications can test the macro's
+value to determine which of the extension's APIs the implementation supports.
+
+[%header,cols="1,5"]
+|===
+|Value |Description
+|1   |Initial extension version.  Base features are supported.
+|===
+
+
+=== New Section 4.7.9.1 annotated_ptr properties
+
+Below is a list of compile-time-constant properties supported with `annotated_ptr`.
+
+```c++
+namespace sycl::ext::oneapi::experimental {
+struct align_key {
+  template<int K>
+  using value_t = property_value<align, std::integral_constant<int, K>>;
+};
+
+template<int K>
+inline constexpr align::value_t<K> align;
+
+template<>
+struct is_property_key<align_key> : std::true_type {};
+
+template<typename SyclObjectT>
+struct is_property_key_of<align_key, SyclObjectT> : std::true_type {};
+
+} // namespace experimental::oneapi::ext::sycl
+```
+--
+[options="header"]
+|====
+| Property | Description
+|`align`
```

Comment:
In addition to an alignment property that takes an integer value as a parameter, I think we'll also want a "runtime_aligment" or "runtime_aligned" property that the user can use to tell the compiler that the pointer has whatever alignment the runtime gave it when allocated (e.g. through a USM malloc call).  The user doesn't necessarily know what that alignment is and it might vary from one target to another, but by applying this property they are promising that they haven't modified that alignment through pointer math.  Each compiler can interpret that alignment as a different number based on its own understanding of its runtime's behaviour.

Reply:
Updated, and added this to the current spec. Please review the change.